src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -502,6 +502,10 @@ static int pid_notify_with_fds_internal(
 
                 type = SOCK_SEQPACKET;
                 fd = socket(address.sockaddr.sa.sa_family, type|SOCK_CLOEXEC, 0);
+                if (fd < 0 && ERRNO_IS_NOT_SUPPORTED(errno)) {
+                        type = SOCK_STREAM;
+                        fd = socket(address.sockaddr.sa.sa_family, type|SOCK_CLOEXEC, 0);
+                }
```

Comment:
hmmm, so or the seqpacket and dgram protocols just calling sendmsg() once is unproblematic. But SOCK_STREAM is different because it allows short writes. And given we have no idea how great or crappy the VM manager is I'd really prefer if we'd have a way to handle this.

so there are two ways to handle this: tape over short writes by calling sendmsg() in a loop. this would be safe and robust I guess but needs some care be taken so that we send the fds along only on the first sendmsg().

the other way would be to simply detect the short write and turn it into EMSGSIZE. This would then mirror what happens if you send too long data via SOCK_DGRAM and SOCK_SEQPACKET: if the data doesn't fit in a single datagram we get EMSGSIZE back. For the streams this would then mean: if the data doesn't fit in a single socket buffer, then you get EMSGSIZE back....

Reply:
hmm, or maybe, thinking about it the EMSGSIZE strategy sucks a bit, because it means the receiver will receieve an incomplete data blob, because unlike the SOCK_DGRAM/SOCK_SEQPACKET case we'd end up writing the first part of the message if there's room but not the rest.

hence maybe bite in the sour apple and do a loop around sendmsg(), but suppress all of the control metadata on all but the first sendmgsg() invocations.

Reply:
IOVEC_INCREMENT() is a nice helper that allows you to skip the data in the struct iovec that you already managed to write, btw